### PR TITLE
Replaced CircularBuffer's uses of checked arithmetic (+, -)

### PR DESCRIPTION
**Pull request by 3a4oT**
Replaced CircularBuffer's uses of checked arithmetic (+, -) with `&+` and `&-`. 
PS. This is my first attempt to contribute so I'm open to any suggestions especially I do not understand how I can verify it makes performance better.

### Motivation:

trying to address  #1275

### Modifications:

Replaced checked arithmetic operations (+, -) with unchecked *&-* and  *&+*

### Result:

better performance

### CODE REVIEW

File: Sources/NIO/CircularBuffer.swift [hunk 1]
```diff
@@ -589,7 +589,7 @@ extension CircularBuffer: RangeReplaceableCollection {
             self.removeSubrange(subrange)
         } else {
             var newBuffer: ContiguousArray<Element?> = []
-            let neededNewCapacity = self.count + newElements.count - subrangeCount + 1 /* always one spare */
+            let neededNewCapacity = self.count &+ newElements.count &- subrangeCount &+ 1 /* always one spare */
```
Comment (weissi) [inline Sources/NIO/CircularBuffer.swift]: I think in this case we probably want the checks as we're growing. I mean we will never be able to get to that many elements but I think it'd still be the right thing to do.

File: Sources/_NIO1APIShims/NIO1APIShims.swift [hunk 2]
```diff
@@ -395,12 +395,12 @@ extension CircularBuffer {
 
     @available(*, deprecated, message: "please use CircularBuffer.Index instead of Int")
     public func index(after: Int) -> Int {
-        return after + 1
+        return after &+ 1
```
Comment (weissi) [inline Sources/_NIO1APIShims/NIO1APIShims.swift]: these ones are deprecated, so I think we might want to take this out

File: Sources/_NIO1APIShims/NIO1APIShims.swift [hunk 3]
```diff
@@ -395,12 +395,12 @@ extension CircularBuffer {
...
     }
 
     @available(*, deprecated, message: "please use CircularBuffer.Index instead of Int")
     public func index(before: Int) -> Int {
-        return before - 1
+        return before &- 1
```
Comment (weissi) [inline Sources/_NIO1APIShims/NIO1APIShims.swift]: these ones are deprecated, so I think we might want to take this out

File: Sources/NIO/CircularBuffer.swift (hunk 1 above)
Comment (3a4oT) [reply Sources/NIO/CircularBuffer.swift]: fixed in a follow-up commit. Thanks

File: Sources/NIO/CircularBuffer.swift [hunk 4]
```diff
@@ -227,7 +227,7 @@ extension CircularBuffer: RandomAccessCollection {
     /// Returns the index offset by `distance` from `index`.
     @inlinable
     public func index(_ i: Index, offsetBy distance: Int) -> Index {
-        return .init(backingIndex: (i.backingIndex + distance) & self.mask,
+        return .init(backingIndex: (i.backingIndex &+ distance) & self.mask,
```
Comment (glbrntt) [inline Sources/NIO/CircularBuffer.swift]: I don't think this one can be left unchecked since `distance` be arbitrarily large. cc @weissi

Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: @glbrntt I _think_ it's undefined behaviour if you ask for an OOB index, let me check the docs.

Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: @glbrntt yes, quoting the [docs](https://developer.apple.com/documentation/swift/bidirectionalcollection/3024450-index)

> The value passed as distance must not offset i beyond the bounds of the collection.

Comment (weissi) [reply Sources/NIO/CircularBuffer.swift]: so I think we're fine here.

Comment (glbrntt) [reply Sources/NIO/CircularBuffer.swift]: ah okay, sounds good to me

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Pull request merged**